src/main/java/com/feedzai/commons/sql/abstraction/batch/AbstractBatch.java

```diff
@@ -305,8 +338,22 @@ public void flush(boolean sync) {
      *
      * @param entries The entries that are pending to be persisted.
      */
-    public abstract void onFlushFailure(BatchEntry[] entries); {
-        // NO-OP.
+    public void onFlushFailure(BatchEntry[] entries) {
+        if (failureListener.isPresent()) {
+            final FailureListener listener = this.failureListener.get();
+            final Set<Map<String, Serializable>> failedEvents = new HashSet<>();
+            for (final BatchEntry entry : entries) {
+                final Map<String, Object> eventMap = entry.getEntityEntry().getMap();
+                final Map<String, Serializable> event = Maps.newHashMapWithExpectedSize(eventMap.size());
+
+                // do not change this to collect(Collectors.toMap()) since it throws an NPE when the value is null
+                eventMap.keySet().
+                        forEach(name -> event.put(name, (Serializable) eventMap.get(name)));
```

Comment:
Can't be done as simply as @marcobjorge is suggesting, because the generic types of the maps are not exactly the same. eventMap's values are Object and event's values are declared as Serializable.